sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,212 @@
+
+            foreach (KeyValuePair<string, string> attribute in ScopeAttributes)
+            {
+                scope.AddAttribute(attribute.Key, attribute.Value);
+            }
+
+            scope.Start();
+
+            try
+            {
+                return await UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+        }
+
+        private async ValueTask<Response> UpdateStateAsync(bool async, CancellationToken cancellationToken)
+        {
+            var state = await _operation.UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
+
+            RawResponse = state.RawResponse;
+
+            if (state.Succeeded == true)
+            {
+                Value = state.Value;
+                HasCompleted = true;
+            }
+            else if (state.Succeeded == false)
+            {
+                _operationFailedException = state.OperationFailedException ?? (async
+                    ? await _diagnostics.CreateRequestFailedExceptionAsync(state.RawResponse).ConfigureAwait(false)
+                    : _diagnostics.CreateRequestFailedException(state.RawResponse));
+                HasCompleted = true;
+
+                throw _operationFailedException;
+            }
+
+            return state.RawResponse;
+        }
+
+        private static TimeSpan GetServerDelay(Response response)
+        {
+            if (response.Headers.TryGetValue(RetryAfterMsHeaderName, out string retryAfterValue)
+                || response.Headers.TryGetValue(XRetryAfterMsHeaderName, out retryAfterValue))
+            {
+                if (int.TryParse(retryAfterValue, out int serverDelayInMilliseconds))
+                {
+                    return TimeSpan.FromMilliseconds(serverDelayInMilliseconds);
+                }
+            }
+
+            if (response.Headers.TryGetValue(RetryAfterHeaderName, out retryAfterValue))
+            {
+                if (int.TryParse(retryAfterValue, out int serverDelayInSeconds))
+                {
+                    return TimeSpan.FromSeconds(serverDelayInSeconds);
+                }
+            }
+
+            return TimeSpan.Zero;
+        }
+    }
+
+    internal interface IOperation<TResult>
+    {
+        ValueTask<OperationState<TResult>> UpdateStateAsync(bool async, CancellationToken cancellationToken);
+    }
+
+    internal readonly struct OperationState<TResult>
+    {
+        private OperationState(Response rawResponse, bool? succeeded, TResult value, RequestFailedException operationFailedException)
+        {
+            RawResponse = rawResponse;
+            Succeeded = succeeded;
+            Value = value;
+            OperationFailedException = operationFailedException;
+        }
+
+        public Response RawResponse { get; }
+
+        public bool? Succeeded { get; }
```

Comment:
Seeing this now (given my previous comment), if you don't actually care about the null state, you could handle that here in the helper e.g.
```csharp
public bool Succeeded => _succeeded.HasValue && _succeeded.Value;
```

Reply:
We need to differentiate because we have three possible states:
- If true: operation has completed successfully.
- If false: operation has completed with errors.
- If null: operation is still running.

Reply:
In general, I get it. But I only remember seeing this property compared to `true` or `false`. Another property could do so as well and at least consolidate the logic I mentioned elsewhere about checking for `HasValue` first to avoid boxing (if the newer compiles aren't doing that automatically for `Nullable<T>` comparisons).

Reply:
> But I only remember seeing this property compared to true or false.

https://github.com/Azure/azure-sdk-for-net/blob/db816eae059233467f93a5d715a0de495d4377dd/sdk/core/Azure.Core/src/Shared/OperationInternal.cs#L139-L152

If it's true, we change the state. If it's false, we change it to something else. If it's null, we keep the state we had previously. We can't achieve the same behavior if we have a non-nullable bool given that we have three possible outcomes.

I'll make `Succeeded` a non-nullable bool and add a new bool called `HasCompleted` so we can avoid the problems mentioned.

Reply:
> If this is what we are aiming for, I second Chris here that this should be documented, commented, etc so people don't have to figure out what each value means

We'll have proper documentation and a migration guide, but developers won't need to understand these values directly. An `OperationState` need to be instantiated by one of its static methods:
- `OperationState<T>.Pending(Response)`
- `OperationState<T>.Success(Response, T)`
- `OperationState<T>.Failure(Response, RequestFailedException)`

Its properties, such as `Succeeded`, are set by these static methods and are used only by the `OperationInternal` class.